### PR TITLE
Enable/Disable distributed training and mixed precision

### PR DESCRIPTION
**Pull request by CCInc**
Adds `distributed` and `mixed_precision` flags to enable/disable those features during training. Let me know your thoughts

### CODE REVIEW

**Comment by CCInc** [comment]
I have finished running my tests. On an rtx 3090 with default training parameters:

Full Precision:
- MinkUNet: 61.2% mIOU, 11 hr training time, 3m21s forward
- SPVCNN: 62.4% mIOU, 11.75hr training time, 3m42s forward

Mixed Precision:
- MinkUNet: 62.4% mIOU, 7.75hr training time, 2m30s forward
- SPVCNN: 62.5% mIOU, 9.4hr training time, 3m19s forward

I think the PR is ready to be reviewed.

**Comment by zhijian-liu** [comment]
Hi @CCInc, thanks for the PR! We can use `torchpack dist-run -np 1` to launch the training with only one process (which might have the same behavior as setting `distributed` to `False`).

File: core/trainers.py [hunk 1]
```diff
@@ -11,20 +12,30 @@
...
+                 seed: int,
+                 mixed_precision: bool = False) -> None:
         self.model = model
         self.criterion = criterion
         self.optimizer = optimizer
         self.scheduler = scheduler
         self.num_workers = num_workers
         self.seed = seed
+        self.mixed_precision = mixed_precision
+        self.loss_scaler = amp.GradScaler(enabled=self.mixed_precision)
```
Comment (zhijian-liu) [inline core/trainers.py:30]: Nit: `loss_scaler` => `scaler`.

Comment (zhijian-liu) [reply core/trainers.py:30]: Remember to save and restore the state of `scaler` (see https://github.com/zhijian-liu/torchpack/blob/master/examples/image-classification/core/trainers.py#L55 and https://github.com/zhijian-liu/torchpack/blob/master/examples/image-classification/core/trainers.py#L62).

File: train.py [hunk 2]
```diff
@@ -83,7 +91,8 @@ def main() -> None:
                                    optimizer=optimizer,
                                    scheduler=scheduler,
                                    num_workers=configs.workers_per_gpu,
-                                   seed=seed)
+                                   seed=seed,
+                                   mixed_precision=configs.mixed_precision)
```
Comment (zhijian-liu) [inline train.py:95]: Nit: `mixed_precision` => `amp_enabled`.

File: core/trainers.py (hunk 1 above)
Comment (zhijian-liu) [reply core/trainers.py:30]: Nit: `loss_scaler` => `scaler`.

**Comment by CCInc** [comment]
Does the `torchpack dist-run -np 1` option still require mpi to be installed? My intention was to alleviate usage problems caused by mpi/mpi4py.

**Comment by zhijian-liu** [comment]
Indeed, it will still require MPI to be installed. A simpler way to avoid MPI is to just remove `dist.init()` at https://github.com/mit-han-lab/spvnas/blob/master/train.py#L23. We don't need to change the other part of the code.

**Comment by CCInc** [comment]
Great, could we keep the `distributed` config param and just have it switch `dist.init()`?

**Comment by zhijian-liu** [comment]
Sure. That works! It would be better to set the default value of `distributed` (in the config file) to `true`.

File: evaluate.py [hunk 3]
```diff
@@ -82,7 +82,8 @@ def main() -> None:
                                    optimizer=optimizer,
                                    scheduler=scheduler,
                                    num_workers=configs.workers_per_gpu,
-                                   seed=configs.train.seed)
+                                   seed=configs.train.seed,
+                                   amp_enabled=False)
```
Comment (zhijian-liu) [inline evaluate.py:86]: This is not necessary as the default value of `amp_enabled` is `False`.

File: README.md [hunk 4]
```diff
@@ -120,6 +120,8 @@ For example, to train the model [SemanticKITTI_val_SPVCNN@30GMACs](https://hanla
 torchpack dist-run -np [num_of_gpus] python train.py configs/semantic_kitti/spvcnn/cr0p5.yaml
 ```
 
+To train the model in a non-distributed enviornment without MPI, i.e. on a single GPU, you may set the `distributed` parameter to `False` in `configs/default.yaml`.
```
Comment (zhijian-liu) [inline README.md:123]: Typo: "enviornment" => "environment".

Comment (zhijian-liu) [reply README.md:123]: I think an easier way is actually to append `--distributed False` to the training command.

**Review by zhijian-liu — APPROVED** [review]
LGTM!

**Pull request merged**